### PR TITLE
Enable sac interactions

### PR DESCRIPTION
**Pull request by ShantelPeters**
Modifications made

### CODE REVIEW

File: src/routes/dashboard/send/+page.svelte [hunk 1]
```diff
@@ -175,19 +177,27 @@ features have been implemented:
      * @function previewPaymentTransaction
      */
     const previewPaymentTransaction = async () => {
+        const destinationAddress = otherDestination ? otherPublicKey : destination;
```
Comment (ElliotFriend) [inline src/routes/dashboard/send/+page.svelte:179]: This is a really great catch! I didn't realize I had duplicated so much for the various `destination` fields. Thanks!

File: src/lib/stores/walletStore.js
Comment (ElliotFriend) [inline src/lib/stores/walletStore.js:1]: There are still leftover changes in this file from an earlier commit you made that need to be reverted. I think there shouldn't be _any_ changes to this file required for the feature you're implementing. Please revert this file to it's current `main` branch state

File: src/lib/stores/contactsStore.js [hunk 2]
```diff
@@ -83,6 +83,15 @@ function createContactsStore() {
                 return false
             }
         },
+
+        /**
+         * Checks if the given address is a contract address
+         * @param {string} address Address to check
+         * @returns {boolean} True if the address is a contract address, false otherwise
+         */
+        isContractAddress: (address) => {
```
Comment (ElliotFriend) [inline src/lib/stores/contactsStore.js:92]: This is a really helpful function to have included. Thanks!

File: src/routes/dashboard/send/+page.svelte [hunk 3]
```diff
@@ -175,19 +177,27 @@ features have been implemented:
...
         let { transaction, network_passphrase } = createAccount
             ? await createCreateAccountTransaction({
                   source: data.publicKey,
-                  destination: otherDestination ? otherPublicKey : destination,
+                  destination: destinationAddress,
                   amount: sendAmount,
                   memo: memo,
               })
+            : StrKey.isValidContract(destinationAddress)
```
Comment (ElliotFriend) [inline src/routes/dashboard/send/+page.svelte]: Let's use the helpful contract check you wrote into the `contactsStore` here. Then, you won't even need to import `StrKey` from the stellar sdk
```suggestion
            : contacts.isContractAddress(destinationAddress)
```

File: src/lib/stellar/transactions.js [hunk 4]
```diff
@@ -1,6 +1,6 @@
-import { TransactionBuilder, Networks, Server, Operation, Asset, Memo } from 'stellar-sdk'
+import { TransactionBuilder, Networks, Operation, Asset, Memo, Contract, xdr, Address, StrKey, SorobanRpc } from 'stellar-sdk'
+import Server from 'stellar-sdk'
```
Comment (ElliotFriend) [inline src/lib/stellar/transactions.js]: This line isn't necessary

File: src/lib/stellar/transactions.js [hunk 5]
```diff
@@ -1,6 +1,6 @@
-import { TransactionBuilder, Networks, Server, Operation, Asset, Memo } from 'stellar-sdk'
+import { TransactionBuilder, Networks, Operation, Asset, Memo, Contract, xdr, Address, StrKey, SorobanRpc } from 'stellar-sdk'
```
Comment (ElliotFriend) [inline src/lib/stellar/transactions.js]: The `rpc` module is the preferred import method for the rpc server from the SDK
```suggestion
import { TransactionBuilder, Networks, Operation, Asset, Memo, Contract, xdr, Address, StrKey, rpc } from '@stellar/stellar-sdk'
```

File: src/lib/stellar/transactions.js [hunk 6]
```diff
@@ -22,6 +23,7 @@ const maxFeePerOperation = '100000'
 const horizonUrl = 'https://horizon-testnet.stellar.org'
 const networkPassphrase = Networks.TESTNET
 const standardTimebounds = 300 // 5 minutes for the user to review/sign/submit
+const rpcUrl = 'https://soroban-rpc.testnet.stellar.org'
```
Comment (ElliotFriend) [inline src/lib/stellar/transactions.js]:
```suggestion
const rpcUrl = 'https://soroban-testnet.stellar.org'
```

File: src/lib/stellar/transactions.js [hunk 7]
```diff
@@ -331,3 +333,52 @@ export async function createPathPaymentStrictReceiveTransaction({
...
+    const server = new Server(horizonUrl);
+    const sourceAccount = await server.loadAccount(source);
+
+    const transaction = new TransactionBuilder(sourceAccount, {
+        networkPassphrase: networkPassphrase,
+        fee: maxFeePerOperation,
+    });
+
+    const [assetCode, assetIssuer] = asset.split(':');
+    const contractId = new Asset(assetCode, assetIssuer).contractId(Networks.TESTNET);
```
Comment (ElliotFriend) [inline src/lib/stellar/transactions.js]: Let's use the already-declared variable from the top of this file
```suggestion
    const contractId = new Asset(assetCode, assetIssuer).contractId(networkPassphrase);
```

File: src/lib/stellar/transactions.js [hunk 8]
```diff
@@ -331,3 +333,52 @@ export async function createPathPaymentStrictReceiveTransaction({
...
+    const transferOp = contract.call(
+        "transfer",
+        Address.fromString(source).toScVal(),
+        Address.fromString(destination).toScVal(),
+        xdr.ScVal.scvI128(new xdr.Int128Parts({
+            lo: xdr.Uint64.fromString(amountBigInt.toString(16).padStart(16, '0').slice(-16)),
+            hi: xdr.Int64.fromString(amountBigInt.toString(16).slice(0, -16) || '0')
+        })),
+        xdr.ScVal.scvSymbol(assetCode)
+    );
```
Comment (ElliotFriend) [inline src/lib/stellar/transactions.js]: This operation is looking pretty good, but might need a little tweaking. The [Token Interface](https://developers.stellar.org/docs/tokens/token-interface#code) defines the `transfer` function as needing `from`, `to`, and `amount` parameters. The `assetCode` one you have at the end is unnecessary. The `amount` value could probably be done more easily (and more straight-forward) by using [the `ScInt` class](https://stellar.github.io/js-stellar-sdk/ScInt.html). You could also use [the `nativeToScVal` function](https://stellar.github.io/js-stellar-sdk/global.html#nativeToScVal) for a lot of these so there's fewer imports at the top, and less complexity around the parameter types (nativeToScVal might also work for the amount, now that I think of it)

This might need some testing/tweaking, but it might look something like this:

```suggestion
    const transferOp = contract.call(
        "transfer",
        nativeToScVal(source, { type: 'address' }),
        nativeToScVal(destination, { type: 'address' }),
        nativeToScVal(amount, { type: 'i128' }),
    );
```

File: src/lib/stellar/transactions.js [hunk 9]
```diff
@@ -331,3 +333,52 @@ export async function createPathPaymentStrictReceiveTransaction({
...
+    transaction.addOperation(transferOp);
+
+    const builtTransaction = transaction.setTimeout(standardTimebounds).build();
+
+    // Simulate the transaction
+    const rpcServer = new SorobanRpc.Server(rpcUrl);
```
Comment (ElliotFriend) [inline src/lib/stellar/transactions.js]:
```suggestion
    const rpcServer = new rpc.Server(rpcUrl);
```

**Review by ElliotFriend — CHANGES_REQUESTED** [review]
Hi, there @ShantelPeters and @akintewe! This is coming along well I've left some comments and suggestions around, but I think you guys are nearly there!

There are also some conflicts and new `main` commits you might have to pull into your feature branch

Thanks for all the hard work!

File: src/lib/stores/contactsStore.js (hunk 2 above)
Comment (ShantelPeters) [reply src/lib/stores/contactsStore.js:92]: You are welcome :)

File: src/routes/dashboard/send/+page.svelte (hunk 3 above)
Comment (ShantelPeters) [reply src/routes/dashboard/send/+page.svelte]: Done

**Comment by ShantelPeters** [comment]
I just made a new Pull Request please check @ElliotFriend

File: package-lock.json
Comment (ElliotFriend) [inline package-lock.json:1]: Let's delete this file, since we use `yarn.lock` for dependency management instead

File: package.json
Comment (ElliotFriend) [inline package.json:1]: the `package.json` file _should_ be fine as-is in the most current version of the `main` branch. I don't think you'll need any changes in this file for your feature.

File: src/lib/stellar/transactions.js [hunk 10]
```diff
@@ -331,3 +334,47 @@ export async function createPathPaymentStrictReceiveTransaction({
...
+ * @param {Object} opts Options object
+ * @param {string} opts.source Public Stellar address to use as the source account of the transaction
+ * @param {string} opts.destination Public Stellar address or contract ID to receive the transfer
+ * @param {string} opts.amount Amount of the asset to transfer
+ * @param {string} opts.asset Asset to be transferred (example: USDC:GBBD47IF6LWK7P7MDEVSCWR7DPUWV3NY3DTQEVFL4NAT4AQH3ZLLFLA5)
+ * @returns {Promise<TransactionResponse>} Object containing the relevant network passphrase and the built transaction envelope in XDR base64 encoding, ready to be signed and submitted
+ */
+export async function createContractTransferTransaction({ source, destination, amount, asset }) {
+    const server = new Server(horizonUrl);
+    const sourceAccount = await server.loadAccount(source);
```
Comment (ElliotFriend) [inline src/lib/stellar/transactions.js]:
```suggestion
    const server = new rpc.Server(rpcUrl);
    const sourceAccount = await server.getAccount(source);
```

We can use the rpc server from the beginning, and then we don't have to make a Horizon request at all.

File: src/lib/stellar/transactions.js [hunk 11]
```diff
@@ -331,3 +334,47 @@ export async function createPathPaymentStrictReceiveTransaction({
...
+        nativeToScVal(destination, { type: 'address' }),
+        nativeToScVal(amount, { type: 'i128' }),
+    );
+    transaction.addOperation(transferOp);
+
+    const builtTransaction = transaction.setTimeout(standardTimebounds).build();
+
+    // Simulate the transaction
+    const rpcServer = new rpc.Server(rpcUrl);
+    const simulatedTx = await rpcServer.prepareTransaction(builtTransaction);
```
Comment (ElliotFriend) [inline src/lib/stellar/transactions.js]:
```suggestion
    const simulatedTx = await server.prepareTransaction(builtTransaction);
```
this first line isn't necessary anymore, if we use the `rpc.Server().getAccount()` function like I mentioned before

File: src/lib/stores/walletStore.js
Comment (ElliotFriend) [inline src/lib/stores/walletStore.js:1]: I think I mentioned this in a previous review, but I believe we don't need _any_ modifications to this file. Please revert the changes to this file.

**Review by ElliotFriend — CHANGES_REQUESTED** [review]
I think you're at the tail-end of this feature! Thanks! I've left some comments throughout addressing some specific things.

I'll reiterate one of them here, so it doesn't get missed. You've got some changes included in the `/src/lib/stores/walletStore.js` file. These changes aren't necessary. This functionality should be accomplished without _any modifications_ to the wallet store file. Please remove those changes from this PR.

Additionally, you've got some conflicting files that need to be resolved. If you pull in all the current changes from the `main` branch, resolve the conflicts, and implement the changes in this review, I think you'll be good to go.

Thanks!

**Comment by ShantelPeters** [comment]
I just made a change please review @ElliotFriend

**Review by ElliotFriend — APPROVED** [review]
Hi @ShantelPeters and @akintewe

This PR looks great! I did make a couple extra commits to resolve some conflicts and fix a couple imports. Thanks so much for all your hard work on this issue! Very well done!
